flang/lib/Lower/OpenMP/OpenMP.cpp

```diff
@@ -2701,7 +2701,42 @@ static void
 genOMP(lower::AbstractConverter &converter, lower::SymMap &symTable,
        semantics::SemanticsContext &semaCtx, lower::pft::Evaluation &eval,
        const parser::OpenMPDeclareMapperConstruct &declareMapperConstruct) {
-  TODO(converter.getCurrentLocation(), "OpenMPDeclareMapperConstruct");
+  fir::FirOpBuilder &firOpBuilder = converter.getFirOpBuilder();
+  lower::StatementContext stmtCtx;
+  const auto &spec =
+      std::get<parser::OmpDeclareMapperSpecifier>(declareMapperConstruct.t);
+  const auto &mapperName{std::get<std::optional<parser::Name>>(spec.t)};
+  const auto &varType{std::get<parser::TypeSpec>(spec.t)};
+  const auto &varName{std::get<parser::Name>(spec.t)};
+  assert(varType.declTypeSpec->category() ==
+             semantics::DeclTypeSpec::Category::TypeDerived &&
+         "Expected derived type");
+
+  std::string mapperNameStr;
+  if (mapperName.has_value())
+    mapperNameStr = mapperName->ToString();
+  else
+    mapperNameStr =
+        "default_" + varType.declTypeSpec->derivedTypeSpec().name().ToString();
+
+  mlir::OpBuilder::InsertPoint insPt = firOpBuilder.saveInsertionPoint();
+  firOpBuilder.setInsertionPointToStart(converter.getModuleOp().getBody());
+  auto mlirType = converter.genType(varType.declTypeSpec->derivedTypeSpec());
+  auto varVal = firOpBuilder.createTemporaryAlloc(
+      converter.getCurrentLocation(), mlirType, varName.ToString());
```

Comment:
Sorry I didn't notice this before.

So far as I understand, this will create the `fir.alloca` and `hlfir.declare` at the beginning of the MLIR module, not nested in any intermediate operation.

How do you intend to lower this to LLVMIR? We would normally nest these in some kind of "function-like" wrapper operation e.g. `func.func` `fir.global` `omp.private` etc. I wonder if the declare mapper operation needs a nested region for this allocation (like we do for `omp.private` and `omp.declare_reduction`).

Reply:
From what I understood when trying to implement this lowering: 
The `declMapperOp's` `parentOp` must have a symbol table as it declares a new symbol, this is why I hoisted it to the `ModuleOp` from the `funcOp`.

The `declMapperOp` also relies on a var definition which is immediately used in it's map clause. I can't put this var inside a region, as then it can't be used in the map clause. As such, this new `alloc` is also moved to just before the `declMapperOp`.

Do you have any suggestion to do this in a different way?

Reply:
Unfortunately I am not very familiar with the target offload parts of OpenMP so I might have missunderstood something here.

OpenMP 5.2 (section 5.8.8) says
> The visibility and accessibility of this declaration are the same as those of a variable declared at the same location in the program.
Therefore I don't think it is right to be pushing this to the module scope in case there are conflicting mappers declared in different scopes in the same module.

If I understand correctly, the `declare mapper` directive declares `var`:
> The variable declared by var is available for use in all map clauses on the directive, and no part of the variable to be mapped is mapped by default.

So it should be treated as though this parse node is a variable declaration so `createTemporaryAlloc` would be appropriate if this is inside of a function, but if it is not, this should probably be a `fir.global` (like if you declared a module- or file-level variable).

Reply:
> Therefore I don't think it is right to be pushing this to the module scope in case there are conflicting mappers declared in different scopes in the same module.

It isn't my intention to move the `declMapperOp` to the module scope, but rather MLIR is forcing me to. If there's an alternative I'm missing where I can keep the declMapperOp within the func/block where it was declared, I'd much rather do that. In case there is no way to do that in MLIR, I can prepend the name of the function to the module scope declaration to prevent conflicts.

> So it should be treated as though this parse node is a variable declaration so `createTemporaryAlloc` would be appropriate if this is inside of a function, but if it is not, this should probably be a `fir.global` (like if you declared a module- or file-level variable).

If we go forward with keeping the `declMapperOp` in the module scope, then I'll alter the `var` declaration to be a `fir.global`.

Reply:
Isn't `declare mapper` providing a specification which is to be used for mapping in the associated regions? The variable in the `declare mapper` has its scope inside the construct since it is only used for specifying how to map. Similar to what @tblah is saying, you could do something like the following.

Assuming %x is the variable that is being mapped by the rules of the declare mapper:
```
omp.declare_mapper @default_my_type {
^bb0(%[[VAR:.*]]: [[TYPE]]):
  %v1 = omp.bounds
  %v2 = omp.map.info var_ptr(%[[VAR:.*]])
  omp.yield %v2
}

%x_map = omp.map.info var_ptr(%x) decl_map(@default_my_type)
omp.target_enter_data   map_entries(%x_map)
```

or 

```
omp.declare_mapper @default_my_type {
^bb0(%[[VAR:.*]]: [[TYPE]]):
  %v1 = omp.bounds
  %v2 = omp.map.info var_ptr(%[[VAR:.*]])
  omp.yield %v2
}
omp.target_enter_data   map_entries(%x : @default_my_type)
```

The alternative would be to not have a `declare mapper` construct and apply the declare mapper during lowering itself. But this would be early lowering and we wont be able to share code with Clang as well.

Reply:
From what I understand, declare mapper directive doesn't simply apply those mapping rules as is to `%x` in your example, or at least not in the plain manner suggested.

A new mapper function is emitted, and a pointer to this function is passed on to the runtime kernel call for target. The OMPRTL takes offload_mappers as the last argument, which is an array of pointers to the user defined mapper funcs.

If we go with the approach you are suggesting above, we would have to decouple the `map clause` from the declare mapper directive and have floating `mapInfoOps` in its nested region. This seems like an off approach to me. Also, we would need to handle this as a special case everywhere we handle regular map clauses, including the several lowering stages and passes like `mapInfoFinalization`.

If we really want to enforce the scoping rule for the declare mapper `var`, perhaps we can do so in the verifier to check that `%var` has a single use, i.e, the `declMapperOp`. Let me know if that sounds reasonable and we can leave the `%var` declaration just above the `declMapperOp`.

As for the placement of the `declMapperOp` at the Module level or Func level, I am undecided. The `declMapperOp` is in some sense akin to a function declaration and will be lowered into one eventually in the llvm IR. We're just delaying that process until we lower to llvm IR, as it saves us unnecessary work, and also we can share the lowering between Clang and MLIR through the OMPIRBUilder.

With that being said, as we are not planning on expanding it into a function at the MLIR level, I am not opposed to leaving it inside the func it's defined in to retain scoping rules. We can do so by changing the `declMapperOp` to no longer declare a symbol for itself, to prevent MLIR from complaining about `ModuleOp` not being it's parent.
This would mean that later when lowering map clauses, instead of doing a sym_table lookup for the mapper, we would have to scan the code for `declMapperOps` instead and match. This may be slightly inefficient, but I don't think should be too much of an issue.

Let me know what are your thoughts on this and how you'd like me to proceed :)

Reply:
Looks like @skatrak is also thinking similarly, see his comment in https://github.com/llvm/llvm-project/pull/117045#discussion_r1858907868

> A new mapper function is emitted, and a pointer to this function is passed on to the runtime kernel call for target. The OMPRTL takes offload_mappers as the last argument, which is an array of pointers to the user defined mapper funcs.

If this is the way it is implemented in Clang, I think you can do it this way by creating the mapper function from the MLIR region-based declare mapper specification operation.

Reply:
@agozillon Do you have any insight or suggestion to handling the declMapperOp map clause one way or the other?

Reply:
I don't have much insight into how the (soon to be) declare mapper portion of the OMPIRBuilder or omp.private/declare_reduction work unfortunately, and I am nowhere near an MLIR guru, still at a beginner level I'd wager, so take my comment with a grain of salt :-) 

I do think it would be ideal to maintain the correct scoping if we can, as the specification seems to be quite clear on it, although, I imagine if there is significant difficulties in doing so at an MLIR level as long as semantically it is the same to a user (i.e. no issues / bugs), then the specification doesn't really have room to complain as we're in the land of implementation details. However, if we can maintain the scoping as it seems to be suggested is possible, that'd be awesome :-)

Otherwise, I do quite like the lambda-esque approach conceptually as it does seem to achieve the desired results as far as scoping goes, and it is tidy in the sense that it can be shared across multiple mappings of the variable/type the mapper has been declared for quite trivially and follows the whole concept that declare mapper is effectively creating a specification/template of the desired mapping. Saying that, I am not against the simpler method either, I am not sure which would be easier to perform transformations/optimisations on if we ever wished to. 

As far as the lowering goes, I don't think it'd be too difficult in either case. The main pass that does things with map information is as you've mentioned MapInfoFinalization, which I believe should only need minor tweaks, primarily to recognize declare mappers as something that contains map.info that needs walked over, you would also likely have to teach it to add any new maps it generates to the yield operation as well. So some tedium but nothing too egregious I think, and more than happy to help with it. I do not think there's anything incredibly complex as far as the OpenMP -> OpenMP dialect translation that occurs when we lower from FIR to LLVM dialect goes as far as the map info, so it'd likely just be a case of mimic'ng how other OMP operations that follow the same format work at that stage. For MLIR -> LLVM-IR, you'd likely still have to do a lot of the same legwork in either case with some differences I would imagine, as at the end of the day the map info and bounds information is just an information vehicle, it'll have to be converted into whatever the declare mapper portion of the OMPIRBuilder desires, which won't be the exact same as what use_device_addr does or what map does. 

But as with anything complex, there will most definitely be issues and unforeseen things you'll likely run into and I am not the soldier in the trenches so to speak, so I imagine you'll be able to assess the difficulties better than I can :-)  

P.S. I would love to see what Clang emits for a declare mapper in the IR if you have a spare blob of IR handy or a small sample program I can compile! Might help to give some further insight into what the end result for us will be (at least it'll help me in any case, will give me a better idea of what the expectations for what the MLIR -> LLVM-IR portion will need to do).

Reply:
> Also, I'm not sure what we stand to gain from this approach. 

In the approach that you are proposing, there is a variable and map.info that is created in the function scope with an alloca. This is not correct since the existence of a declare mapper does not cause the creation of a variable or a mapping.

Worst case, 
-> this can appear to be aliasing with other memory locations and reduce performance. 
-> increase live-ins of openmp regions when outlined

> Every other instance of where a directive has a mapClause associated with it, does so by creating the map related Ops just above it. Why are we nesting them in case of declMapperOp?

Because it is a declaration. It will only have an effect when it is used either via its usage in implicit/default mapping of the derived type for which it was created or when it is directly referred to by a map(mapper...) clause.

I think you have a reasonable point here because of the following:
```
1. If a DECLARE MAPPER directive is not specified for a type DT, a predefined mapper exists for type DT as if the type DT had appeared in the directive as follows:

!$OMP DECLARE MAPPER (DT :: var) MAP (TOFROM: var)
2. If a variable is not a scalar then it is treated as if it had appeared in a map clause with a map-type of tofrom. Which is effectively equivalent to the following and extending declare mapper for non-derived types:
!$OMP DECLARE MAPPER (T :: var) MAP (TOFROM: var)
```
We can thus argue that all mapping should also generate declarations that are ultimately all handled in a separate pass in flang/mlir. May be 1 and 2 are trivial and that is why it is not done that way.

Reply:
Besides the representation, we should also clarify 
-> where we create the map_entries for the relevant operations (like target) for which the declare mapper implicitly applies
-> where the composition of map-types (map-type decay) from the map clause of declare mapper and the map clause of the relevant operation (like target) happens

Note: All these are for discussion and should be treated as suggestions.

Reply:
> ```
> 1. If a DECLARE MAPPER directive is not specified for a type DT, a predefined mapper exists for type DT as if the type DT had appeared in the directive as follows:
> 
> !$OMP DECLARE MAPPER (DT :: var) MAP (TOFROM: var)
> 2. If a variable is not a scalar then it is treated as if it had appeared in a map clause with a map-type of tofrom. Which is effectively equivalent to the following and extending declare mapper for non-derived types:
> !$OMP DECLARE MAPPER (T :: var) MAP (TOFROM: var)
> ```

I think the keyword here is likely "as if", so as long as the effects are as described it's reasonable would be my reading, and if we really wanted to be exact about the wording we'd generate/embed our own equivalent pragmas to the above for all default mappings, and then lower them, so not just at the MLIR level. However, saying that I am not against defining a default declare mapper for all cases once it's in place,  it might tidy things up a bit, but it may also be more complicated/trouble than it's worth, in either case I am fine with the approach of defining default declare mappers if we'd like to go down that route :-) 

I'd also love it if whatever implementation we landed on was compatible with the OpenACC implementations documentation/approach to mapping descriptors via runtime calls, as I'd like to move towards that eventually when I have some time to dig into it and see if it's viable for us. I imagine it will be, I just don't know a ton about the region'd approach so hope it wouldn't be prohibitive of this.

Reply:
I have made the required changes to both PRs to update the representation for the DeclMapperOp.

Some changes I've incorporated are:

Inside DeclMapperOp's region I've introduce a new `DeclMapperInfoOp` to which I've attached the `MapClause`. Not having any `MapClause` explicitly associated seemed weird to me, also walking through the region collecting all the `MapInfoOps` for processing in various places in the code base seemed like a bad design to me.

Also, instead of using the block arg, I've created a `temporary alloca` inside the region. This is to save the hassle of binding the block_arg to the symbol, binding the `alloca` is much more straightforward.

We can drop the entire `DeclMapperOp` including the region once it reaches `OpenMPTOLLVMIRTranslation`.

Let me what are your thoughts on this approach :)

Reply:
@kiranchandramohan @tblah @skatrak @agozillon Any thoughts on this approach?

Reply:
You should feel free to proceed with the approval of @agozillon and @skatrak. You can set up a quick call with them and describe your approach and come to a conclusion. You can report the outcomes somewhere in this patch or write in discourse. 

> Inside DeclMapperOp's region I've introduce a new DeclMapperInfoOp to which I've attached the MapClause. Not having any MapClause explicitly associated seemed weird to me, also walking through the region collecting all the MapInfoOps for processing in various places in the code base seemed like a bad design to me.

The idea would generally be to inline the whole declare mapper operation region, replacing the block_arg with the real variable that is going to be mapped. Would an alloca always be created for all kinds of mappings? If not committing to an alloca might mean you have to delete it in some circumstances.

> We can drop the entire DeclMapperOp including the region once it reaches OpenMPTOLLVMIRTranslation.

As in just drop it without using it? Or using it create the `@.omp_mapper._ZTS1T.deep_copy` function for the declare mapper (`#pragma omp declare mapper(deep_copy : T abc) map(abc, abc.ptr[ : abc.buf_size])`) in the C example you gave below.

You have not talked about the following two points.
```
-> where we create the map_entries for the relevant operations (like target) for which the declare mapper implicitly applies. Currently, this is done during lowering. But in this patch you have solely focused on creating the declare mapper.
-> where the composition of map-types (map-type decay) from the map clause of declare mapper and the map clause of the relevant operation (like target) happens
```

Couple of other things that came to my mind :
-> Since declare mappers are in the specification section, it can also occur in modules. We have not added code to propagate it to the module file. If this is not urgent for you, I can fix it sometime.
-> You should test the case where the declare mapper is in the host subroutine
```
subroutine A
type t
end type
declare mapper(t)
contains
subroutine B
!$omp target 
! use a var of type t
!$omp end target
end subroutine
```

Reply:
> You should feel free to proceed with the approval of @agozillon and @skatrak.
Thanks for the go ahead, I'll consult with them before proceeding, and post any relevant findings here.

> > Inside DeclMapperOp's region I've introduce a new DeclMapperInfoOp to which I've attached the MapClause. Not having any MapClause explicitly associated seemed weird to me, also walking through the region collecting all the MapInfoOps for processing in various places in the code base seemed like a bad design to me.
> 
> The idea would generally be to inline the whole declare mapper operation region, replacing the block_arg with the real variable that is going to be mapped. Would an alloca always be created for all kinds of mappings? If not committing to an alloca might mean you have to delete it in some circumstances.

In case of DeclareMapper the mapping is handled in a ad-hoc basis. There is no real variable to be replaced later. The variable simply exists as a dummy to represent the mapping information for other real variables where the mapping would be used. However, this mapping isn't applied directly to those variables either, rather the runtime is notified that a mapper function exists for these variables.

> > We can drop the entire DeclMapperOp including the region once it reaches OpenMPTOLLVMIRTranslation.
> 
> As in just drop it without using it? Or using it create the `@.omp_mapper._ZTS1T.deep_copy` function for the declare mapper (`#pragma omp declare mapper(deep_copy : T abc) map(abc, abc.ptr[ : abc.buf_size])`) in the C example you gave below.
Yes, I mean't each declareMapperOp would be resolved to a function like `@.omp_mapper._ZTS1T.deep_copy`. By dropping it, I mean't it's region won't be separately lowered.

> You have not talked about the following two points.
> 
> ```
> -> where we create the map_entries for the relevant operations (like target) for which the declare mapper implicitly applies. Currently, this is done during lowering. But in this patch you have solely focused on creating the declare mapper.
> -> where the composition of map-types (map-type decay) from the map clause of declare mapper and the map clause of the relevant operation (like target) happens
> ```

I'll address the implicit mapping either in the DeclMapper lowering support for the mapClause or a separate PR soon after that.

The composition of mapClause AFAIK is likely handled in the OMPIRBuilder. If that's not the case, I'll address it in a future patch. 

> Couple of other things that came to my mind : -> Since declare mappers are in the specification section, it can also occur in modules. We have not added code to propagate it to the module file. If this is not urgent for you, I can fix it sometime. 

Please feel free to do this if you're already familiar with what needs to be done, as I'm not. Otherwise, I can add it to my TODO list :)

-> You should test the case where the declare mapper is in the host subroutine
> 
> ```
> subroutine A
> type t
> end type
> declare mapper(t)
> contains
> subroutine B
> !$omp target 
> ! use a var of type t
> !$omp end target
> end subroutine
> ```

Thanks, I'll test this in the mapClause mapper support PR once I have it ready.

Many thanks for the review.